### DISCUSSION TITLE
Azure SAML logout 500 error

### DISCUSSION THREAD

**Discussion by fredo2906** [kimai/kimai#4599] [Questions & Answers] [answered]
Hello,

First, thanks for this nice application. 
I am trying to setup sso with Azure and everything works fine until a user tries to logout, the logout in Azure works OK but the redirect seems failing. I looked around to see if there were similar issues, found a few, tried what was discussed but nothing seems to work.

The 500 error occurs on the redirect after azure logout such as https://time.mydomain.com/auth/saml/logout?SAMLResponse=.....

If I click the button "Take me home" on that page, then click "Logout" again, I really get logged out from Kimai.

Azure is configured like this :

Basic SAML Configuration
Identifier (Entity ID)
https://time.mydomain.com/auth/saml/metadata
Reply URL (Assertion Consumer Service URL)
https://time.mydomain.com/auth/saml/acs
Sign on URL
https://time.mydomain.com/
Relay State (Optional)
Optional
Logout Url (Optional)
https://time.mydomain.com/auth/saml/logout

Attributes & Claims
givenname -> user.givenname
surname -> user.surname
name -> user.userprincipalname
displayname -> user.displayname
emailaddress -> user.mail
Unique User Identifier -> user.userprincipalname

The local.yaml

kimai:
    saml:
        provider: microsoft
        activate: true
        title: Azure AD SAML
        mapping:
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name, kimai: username }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress, kimai: email }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname, kimai: alias }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/displayname, kimai: title }
        roles:
            attribute: Roles
            mapping:
                - { saml: Admins, kimai: ROLE_ADMIN }
                - { saml: Management, kimai: ROLE_TEAMLEAD }
        connection:
            idp:
                entityId: 'https://sts.windows.net/0aabea60-8946-425a-xxx/'
                singleSignOnService:
                    url: 'https://login.microsoftonline.com/0aabea60-8946-425a-ba12-xxx/saml2'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                singleLogoutService:
                    url: 'https://login.microsoftonline.com/0aabea60-8946-425a-ba12-xxx/saml2'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                x509cert: 'long string'
            sp:
                entityId: 'https://time.mydomain.com/auth/saml/metadata'
                assertionConsumerService:
                    url: 'https://time.mydomain.com/auth/saml/acs'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
                singleLogoutService:
                    url: 'https://time.mydomain.com/auth/saml/logout'
                    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
            strict: true
            debug: true
            security:
                requestedAuthnContext: true

Here are the logs I see at the time the error occurs : 

[2024-01-31 05:38:55] [request] INFO: Matched route "saml_logout". {"route":"saml_logout","route_parameters":{"_route":"saml_logout","_controller":"App\\Controller\\Auth\\SamlController::logoutAction"},"request_uri":"https://time.mydomain.com/auth/saml/logout?RelayState=https%3A%2F%2Ftime.mydomain.com%2Fen%2Flogout&SAMLResponse=fZJRa4MwEMe%2FiuQ9JjHqYrDC2MYodBusZQ97GWeMnUMT8SLdx19r6UNh9PGO%2B%2Bd%2BP3IlwtCPeuP3fg7vFkfv0EbrxxX5krWqW65amkKhaCqKgiqTSSqaJuGtaWolLYk%2B7ISddyuSxJxEa8TZrh0GcOHY4klKuaBS7HimpdJZGuci%2ByTRo8XQOQhL8juEETVjoRtsbAc77Tu3pz9jbPzAYA7f7ATJ%2BoXxuMNdOHd%2BRd5enzZvz%2BvXL2VSpdomr5UQJ747KQuRFW0ulUwUtLloFBeiJtHv0DvUi%2FiKzJPTHrBD7WCwqIPR2%2FuXjT7q6HHywRvfk6pcxKZz9HYIEO10EiPVRQwDxofONf6AsbOBcYDaQs6pKtKcpkkGtAaR0NbI2iSQNaoAVrLzzqo8f9E2QJjxunrwjY0%2BoJ%2FtbSZcpvV2NsYiElaV7PpR9t8ZVH8%3D&SigAlg=http%3A%2F%2Fwww.w3.org%2F2001%2F04%2Fxmldsig-more%23rsa-sha256&Signature=xLR39MX9N8n71yX4hGFXZVTLp4ygr6MCFJ5kUnNX9HHvHORo%2FDY7VSV%2FSf33nlSVwrXbQdjssny%2BWRWUWxTqQ%2FWGZYfgbR1b4yzcia%2FXkVj7u1BCHoGRn3uyyGOVmoiahZlyQGXoolJ3JkRLPSJAXvIZCb2XiCVkerJqBQ5KpSo1SnpqBzQz0wSTXp0MI%2FU3K0MXKOwZnb56PlhAbAurpOPXlM3B131PwI1HW3xTb7bHhJ0u6dhCmIMIOFBbPBonBL1%2Br789M3SpXpV4UpHG6m1C09yfHh3hRx7cnfOiniQCpOZ8cdYFyLozlLQY7rdRkSSD%2F9N0sKZaosiDP5LlqA%3D%3D","method":"GET"} []
[2024-01-31 05:38:55] [security] INFO: Remember-me cookie accepted. [] []
[2024-01-31 05:38:55] [security] INFO: Authenticator successful! {"token":{"Symfony\\Component\\Security\\Core\\Authentication\\Token\\RememberMeToken":"RememberMeToken(user=\"kimai88@mydomain.com\", roles=\"ROLE_USER\")"},"authenticator":"Symfony\\Component\\Security\\Http\\Authenticator\\RememberMeAuthenticator"} []
[2024-01-31 05:38:55] [request] CRITICAL: Uncaught PHP Exception RuntimeException: "You must configure the logout path in your firewall." at SamlController.php line 96 {"exception":"[object] (RuntimeException(code: 0): You must configure the logout path in your firewall. at /var/www/html/kimai/src/Controller/Auth/SamlController.php:96)"} []

I hope someone can help fixing the issue, or at least giving some direction.

Thanks a lot in advance.

**Comment by fredo2906** [comment, marked as answer]
After some digging I got it working.

The trick was to change to "APP_ENV=dev" in .env and run "sudo composer install --optimize-autoloader" so that we can see a little bit more of what is actually failing/happening.

The issue was that, when Azure redirect to Kimai for the logout, Kimai checks if the csrf is valid and since it is not Kimai is complaining. The other issue I had is that it is needed to set the correct path in security.yaml as per 
        secured_area:
            logout:
                path: /auth/saml/logout
                target: /en/login
                enable_csrf: false
                invalidate_session: true

Now, the logout redirect correctly to /en/login (i verified the session is correctly gone in both Azure and local, no more access possible).

Thanks.

**Reply by kevinpapst** [reply]
Thanks for sharing your solution!

In my world, an app should not perform a SSO Logout, but only a logout for itself.
To me it would feel very unexpected (to say at least), if I log out of Kimai and suddenly I can't access my emails / usual work environment. 
That might even reduce security, when people about this behavior and simply close the Kimai tab instead of logging out.

I might be wrong though as "logout" is not a feature that I use often...

**Reply by fredo2906** [reply]
Yes, sorry for my wrong explanation. I am still logged in azure other applications but not in Kimai anymore. The SSO part was working fine, only the  security "firewall" setting was not set as it should be., hence the error and the app was not redirected to the login page.

below is what actually fixed the issue, changes in security.yaml

secured_area:
    logout:
        path: /auth/saml/logout
        target: /en/login
        enable_csrf: false
        invalidate_session: true

**Reply by kevinpapst** [reply]
If you are only trying to logout of Kimai, do NOT configure a SSO logout.
Leave it empty and it will work without you changing core files, which will lead to merge problems during the next update.

**Reply by snapet** [reply]
Hi Kevin,

In my setup I login with my admin account to administer Kimai and with my regular account when using it for actual time tracking. In that scenario I suppose it makes sense to logout. I am glad I found the above solution but am worried a bit that I will need to make the changes every time I update Kimai.
**Can those security settings be captured in the local.yaml file also?**
If not I will happily live with making the changes or merging.

Peter

**Comment by kevinpapst** [comment]
Re-opening, as the answer gets lost otherwise. Discussions should be marked as answered.

**Comment by fredo2906** [comment]
I didnt even know it was possible, for me it was a set. Thanks for the
advice!

Le mer. 31 janv. 2024 à 18:49, Kevin Papst ***@***.***> a
écrit :

**Comment by bykof** [comment]
After an upgrade to 2.12.0 and using Traefik instead of NGINX I got the error: `An authentication exception occurred.` all the time with Azure SAML, which worked without problems before the update.

Setting the env variable: `APP_ENV=dev` fixed the problem o.O

**Reply by kevinpapst** [reply]
Nothing changed regarding the logout
